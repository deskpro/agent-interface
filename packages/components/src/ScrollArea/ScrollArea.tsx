import * as React from "react";
import classNames from "classnames";
import Scrollbars from "react-custom-scrollbars";
import * as debounce from "lodash/debounce";

export type ScrollAreaProps = {
  className?: string;
  contentClassName?: string;
  horizontal?: boolean;
  vertical?: boolean;
  threshold?: number;
  onTresholdReach?(): void;
  style?: object;
};

const ScrollArea: React.FC<ScrollAreaProps & { scrollRef?: React.Ref<{}> }> = ({
  className,
  contentClassName,
  threshold,
  onTresholdReach,
  scrollRef,
  style,
  children
}) => {
  const scrollCallback = React.useCallback(
    debounce(values => {
      if (!threshold || !onTresholdReach) {
        return;
      }

      if (
        (threshold <= 1 && values.top >= threshold) ||
        (threshold > 1 && values.clientHeight * values.top >= threshold)
      ) {
        onTresholdReach();
      }
    }, 250),
    [threshold]
  );

  return (
    <Scrollbars
      className={classNames("dp-ScrollArea", className)}
      ref={scrollRef}
      // autoHeight
      // hideTracksWhenNotNeeded
      style={style}
      onScrollFrame={scrollCallback}
      renderTrackVertical={(p: object) => (
        <div {...p} className="dp-ScrollArea-trackVertical" />
      )}
      renderThumbVertical={(p: object) => (
        <div {...p} className="dp-ScrollArea-thumbVertical" />
      )}
      renderTrackHorizontal={(p: object) => (
        <div {...p} className="dp-ScrollArea-trackHorizontal" />
      )}
      renderThumbHorizontal={(p: object) => (
        <div {...p} className="dp-ScrollArea-thumbHorizontal" />
      )}
      renderView={(p: object) => (
        <div
          {...p}
          className={classNames("dp-ScrollArea-view", contentClassName)}
        />
      )}
    >
      {children}
    </Scrollbars>
  );
};

export default React.forwardRef((props, ref) => (
  <ScrollArea {...props} scrollRef={ref} />
)) as React.FC<ScrollAreaProps>;
