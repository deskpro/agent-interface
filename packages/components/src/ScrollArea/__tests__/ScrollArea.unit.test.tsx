import * as React from "react";
import { render } from "react-testing-library";

import ScrollArea from "../ScrollArea";

describe("<ScrollArea />", () => {
  it("should match its snapshot", () => {
    const { getByTestId } = render(
      <ScrollArea data-testid="scrollarea-test">
        <div>Content</div>
      </ScrollArea>
    );
    expect(getByTestId("scrollarea-test")).toMatchSnapshot();
  });
});
