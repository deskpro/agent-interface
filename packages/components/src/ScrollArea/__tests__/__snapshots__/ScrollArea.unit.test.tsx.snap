// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScrollArea /> should match its snapshot 1`] = `
<div
  class="dp-ScrollArea"
  data-testid="scrollarea-test"
  style="position: relative; overflow: hidden; width: 100%; height: 100%;"
>
  <div
    class="dp-ScrollArea-view"
    style="position: absolute; top: 0px; left: 0px; right: 0px; bottom: 0px; overflow: scroll; margin-right: 0px; margin-bottom: 0px;"
  >
    <div>
      Content
    </div>
  </div>
  <div
    class="dp-ScrollArea-trackHorizontal"
    style="position: absolute; height: 6px; transition: opacity 200ms; opacity: 0; display: none;"
  >
    <div
      class="dp-ScrollArea-thumbHorizontal"
      style="position: relative; display: block; height: 100%;"
    />
  </div>
  <div
    class="dp-ScrollArea-trackVertical"
    style="position: absolute; width: 6px; transition: opacity 200ms; opacity: 0; display: none;"
  >
    <div
      class="dp-ScrollArea-thumbVertical"
      style="position: relative; display: block; width: 100%;"
    />
  </div>
</div>
`;
